mixer/cmd/build.go

```diff
@@ -110,6 +112,144 @@ var buildBundlesCmd = &cobra.Command{
 	},
 }
 
+// buildOldFormatCmd is used to build the final version in the current format (the +10)
+// and ready the mix state for the new (+20) version to be built by a (possibly) newer mixer
+var buildOldFormatCmd = &cobra.Command{
+	Use:   "format-old",
+	Short: "Build the last version in the current format for the format bump",
+	Long:  `Build the last version in the current format for the format bump`,
+	Run: func(cmd *cobra.Command, args []string) {
+		b, err := builder.NewFromConfig(config)
+		if err != nil {
+			fail(err)
+		}
+
+		// Update the mixversion just in case the user did not pass --increment
+		// This must be the +20 to write the new format data files even though we
+		// will build a +10 from the same content
+		for i := 0; i < 2; i++ {
+			if err = b.UpdateMixVer(); err != nil {
+				failf("Couldn't update Mix Version")
+			}
+			// Must re-read the builder or UpdateMixVer() will be a noop each iteration
+			b, err = builder.NewFromConfig(config)
+			if err != nil {
+				fail(err)
+			}
+		}
+
+		// Set format to format+1 so that the format file inserted into the
+		// update content is the new one
+		if err = builder.UpdateFormatVersion(b, buildFlags.format); err != nil {
+			fail(err)
+		}
+		setWorkers(b)
+
+		// Re-read builder.conf after updating format/mixver
+		b, err = builder.NewFromConfig(config)
+		if err != nil {
+			fail(err)
+		}
+
+		fmt.Println(" Backing up full groups.ini")
+		// Back up groups.ini in case we have deprecated bundles to delete
+		if err = builder.CopyFullGroupsINI(b); err != nil {
+			fail(err)
+		}
+		// Fill this in w/Update bundle definitions
+		// if err := UpdateBudlesForFormatBump(); err != nil {...}
+
+		// Build the +20 (first build in new format) bundles
+		if err = buildBundles(b, buildFlags.noSigning); err != nil {
+			fail(err)
+		}
+
+		fmt.Println("Backing up trimmed groups.ini for future build")
+		// Copy the new trimmed groups.ini to use for the +20 build which will
+		// not have manifest created for fully deleted bundles
+		if err = builder.CopyTrimmedGroupsINI(b); err != nil {
+			fail(err)
+		}
+
+		// Copy +20 chroots to +10 so we can build last formatN build with the
+		// same content
+		prevVersion, err := strconv.Atoi(b.MixVer)
+		if err != nil {
+			fail(err)
+		}
+		prevVersion -= 10
+		source := filepath.Join(b.Config.Builder.ServerStateDir, "image", b.MixVer)
+		dest := filepath.Join(b.Config.Builder.ServerStateDir, "image", strconv.Itoa(prevVersion))
+		fmt.Println(" Copying +20 chroots to +10 chroots")
+		if err = helpers.RunCommandSilent("cp", "-al", source, dest); err != nil {
```

Comment:
It's easy to do an exec like this, but could you achieve the same thing natively with `os.Link()` in an `ioutil.ReadDir()` loop? Or would you be losing the effects of the `--archive` you're passing to `cp`?

Reply:
cp handles cross device copies. Of course we don't usually have extra disks mounted here, but we could.

A os.Link() would certainly be much faster.

Reply:
The cp -al is what is currently done in the format bump flow and as icarus pointed out, handles copying over device boundaries better.